File: 1833.maximum-ice-cream-bars/solution.go
func maxIceCream(costs []int, coins int) int {
    const maxCost = 100_000
    var count [maxCost+1]int

    for _, c := range costs {
        count[c]++
    }

    res := 0
    for c, n := range count {
        if c > coins {
            break
        } 
        if n == 0 {
            continue
        }
        b := min(n, coins/c)
        res += b
        coins -= b*c
    }

    return res
}

File: 1833.maximum-ice-cream-bars/solution.kt
class Solution {
    fun maxIceCream(costs: IntArray, coins: Int): Int {
        val maxCost = costs.max()
        val counts = IntArray(maxCost + 1)

        for (cost in costs) {
            counts[cost]++
        }

        var res = 0
        var rest = coins

        for (cost in 1..maxCost) {
            if (cost > rest) {
                break
            }
            if (counts[cost] == 0) {
                continue
            }
            val bars = minOf(counts[cost], rest / cost)
            res += bars
            rest -= bars * cost
        }

        return res
    }
}